File: go/bittorrent/krpc.go
// SPDX-License-Identifier: MIT

package bittorrent

import "fmt"

// KrpcType is the KRPC message kind (the "y" field).
type KrpcType int

const (
	KrpcQuery KrpcType = iota
	KrpcResponse
	KrpcError
)

// KrpcMessage is a KRPC message (BEP-5): a bencoded dict with "t" (transaction id),
// "y" (q/r/e), and one of a query ("q"+"a"), response ("r"), or error ("e").
type KrpcMessage struct {
	TransactionID []byte
	Type          KrpcType
	Method        string // query: the "q" method name
	Arguments     *BDict // query: "a"
	Response      *BDict // response: "r"
	ErrorCode     int64  // error: e[0]
	ErrorMessage  string // error: e[1]
}

// Encode serializes the message to canonical bencode.
func (m KrpcMessage) Encode() ([]byte, error) {
	d := NewBDict()
	if err := d.Add("t", BStr(m.TransactionID)); err != nil {
		return nil, err
	}
	switch m.Type {
	case KrpcQuery:
		_ = d.Add("y", BStr("q"))
		_ = d.Add("q", BStr(m.Method))
		if m.Arguments != nil {
			_ = d.Add("a", m.Arguments)
		} else {
			_ = d.Add("a", NewBDict())
		}
	case KrpcResponse:
		_ = d.Add("y", BStr("r"))
		if m.Response != nil {
			_ = d.Add("r", m.Response)
		} else {
			_ = d.Add("r", NewBDict())
		}
	case KrpcError:
		_ = d.Add("y", BStr("e"))
		e := BList{BInt(m.ErrorCode), BStr(m.ErrorMessage)}
		_ = d.Add("e", e)
	default:
		return nil, fmt.Errorf("unknown KRPC type %d", m.Type)
	}
	return Encode(d), nil
}

// DecodeKrpc parses a KRPC message.
func DecodeKrpc(data []byte) (KrpcMessage, error) {
	var m KrpcMessage
	v, err := Decode(data)
	if err != nil {
		return m, err
	}
	d, err := AsDict(v)
	if err != nil {
		return m, err
	}

	tVal, ok := d.Get("t")
	if !ok {
		return m, fmt.Errorf("KRPC message has no 't'")
	}
	if m.TransactionID, err = AsBytes(tVal); err != nil {
		return m, err
	}

	yVal, ok := d.Get("y")
	if !ok {
		return m, fmt.Errorf("KRPC message has no 'y'")
	}
	y, err := AsText(yVal)
	if err != nil {
		return m, err
	}

	switch y {
	case "q":
		m.Type = KrpcQuery
		qVal, ok := d.Get("q")
		if !ok {
			return m, fmt.Errorf("KRPC query has no 'q'")
		}
		if m.Method, err = AsText(qVal); err != nil {
			return m, err
		}
		if aVal, ok := d.Get("a"); ok {
			m.Arguments, _ = AsDict(aVal)
		}
	case "r":
		m.Type = KrpcResponse
		if rVal, ok := d.Get("r"); ok {
			m.Response, _ = AsDict(rVal)
		}
	case "e":
		m.Type = KrpcError
		if eVal, ok := d.Get("e"); ok {
			if list, err := AsList(eVal); err == nil && len(list) >= 2 {
				m.ErrorCode, _ = AsInt(list[0])
				m.ErrorMessage, _ = AsText(list[1])
			}
		}
	default:
		return m, fmt.Errorf("unknown KRPC y=%q", y)
	}
	return m, nil
}

File: swift/Sources/AetherNetBitTorrent/Krpc.swift
// SPDX-License-Identifier: MIT

import Foundation

/// The KRPC message kind (the "y" field).
public enum KrpcType: Equatable {
    case query
    case response
    case error
}

/// A KRPC message (BEP-5): a bencoded dict with "t" (transaction id), "y" (q/r/e),
/// and one of a query ("q"+"a"), response ("r"), or error ("e").
public struct KrpcMessage {
    public var transactionID: [UInt8]
    public var type: KrpcType
    public var method: String        // query: the "q" method name
    public var arguments: BDict?     // query: "a"
    public var response: BDict?      // response: "r"
    public var errorCode: Int64      // error: e[0]
    public var errorMessage: String  // error: e[1]

    public init(transactionID: [UInt8],
                type: KrpcType,
                method: String = "",
                arguments: BDict? = nil,
                response: BDict? = nil,
                errorCode: Int64 = 0,
                errorMessage: String = "") {
        self.transactionID = transactionID
        self.type = type
        self.method = method
        self.arguments = arguments
        self.response = response
        self.errorCode = errorCode
        self.errorMessage = errorMessage
    }

    /// Serializes the message to canonical bencode.
    public func encode() throws -> [UInt8] {
        let d = BDict()
        try d.add("t", .bytes(transactionID))
        switch type {
        case .query:
            try d.add("y", .text("q"))
            try d.add("q", .text(method))
            try d.add("a", .dict(arguments ?? BDict()))
        case .response:
            try d.add("y", .text("r"))
            try d.add("r", .dict(response ?? BDict()))
        case .error:
            try d.add("y", .text("e"))
            try d.add("e", .list([.int(errorCode), .text(errorMessage)]))
        }
        return bencodeEncode(.dict(d))
    }
}

public enum KrpcError: Error, Equatable {
    case malformed(String)
}

/// Parses a KRPC message.
public func decodeKrpc(_ data: [UInt8]) throws -> KrpcMessage {
    let v = try bencodeDecode(data)
    let d = try v.dictValue()

    guard let tVal = d.get("t") else {
        throw KrpcError.malformed("KRPC message has no 't'")
    }
    let transactionID = try tVal.bytesValue()

    guard let yVal = d.get("y") else {
        throw KrpcError.malformed("KRPC message has no 'y'")
    }
    let y = try yVal.textValue()

    switch y {
    case "q":
        guard let qVal = d.get("q") else {
            throw KrpcError.malformed("KRPC query has no 'q'")
        }
        let method = try qVal.textValue()
        var args: BDict? = nil
        if let aVal = d.get("a") { args = try? aVal.dictValue() }
        return KrpcMessage(transactionID: transactionID, type: .query, method: method, arguments: args)
    case "r":
        var resp: BDict? = nil
        if let rVal = d.get("r") { resp = try? rVal.dictValue() }
        return KrpcMessage(transactionID: transactionID, type: .response, response: resp)
    case "e":
        var code: Int64 = 0
        var message = ""
        if let eVal = d.get("e"), let list = try? eVal.listValue(), list.count >= 2 {
            code = (try? list[0].intValue()) ?? 0
            message = (try? list[1].textValue()) ?? ""
        }
        return KrpcMessage(transactionID: transactionID, type: .error, errorCode: code, errorMessage: message)
    default:
        throw KrpcError.malformed("unknown KRPC y=\(y)")
    }
}
